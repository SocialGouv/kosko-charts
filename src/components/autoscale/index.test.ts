//

import { Deployment } from "kubernetes-models/apps/v1/Deployment";

beforeEach(() => {
  jest.resetModules();
});

test("should create an autoscale for a deployment", async () => {
  jest.doMock(
    "@socialgouv/kosko-charts/environments/gitlab",
    () => gitlabModuleMock
  );

  const deployment = new Deployment({
    metadata: {
      name: "foo",
      namespace: "foo-namespace",
    },
  });

  const { createAutoscale } = await import("./index");

  const hpa = createAutoscale(deployment);

  expect(hpa).toMatchSnapshot();
});

test("should create an autoscale with custom spec", async () => {
  jest.doMock(
    "@socialgouv/kosko-charts/environments/gitlab",
    () => gitlabModuleMock
  );

  const deployment = new Deployment({
    metadata: {
      name: "foo",
      namespace: "foo-namespace",
    },
  });

  const { createAutoscale } = await import("./index");

  const hpa = createAutoscale(deployment, {
    maxReplicas: 4,
    metrics: [
      {
        resource: {
          name: "cpu",
          target: {
            averageUtilization: 50,
            type: "Utilization",
          },
        },
        type: "Resource",
      },
    ],
    minReplicas: 2,
  });

  expect(hpa).toMatchSnapshot();
});

test("fails if without metadata on target", async () => {
  const deployment = new Deployment({});

  const { createAutoscale } = await import("./index");
  expect(() => createAutoscale(deployment)).toThrowErrorMatchingInlineSnapshot(
    `"Missing metadata.name on target"`
  );
});

test("fails if without name on target", async () => {
  const deployment = new Deployment({
    metadata: {
      namespace: "foo-namespace",
    },
  });

  const { createAutoscale } = await import("./index");
  expect(() => createAutoscale(deployment)).toThrowErrorMatchingInlineSnapshot(
    `"Missing metadata.name on target"`
  );
});

test("fails if without namespace on target", async () => {
  const deployment = new Deployment({ metadata: { name: "foo" } });

  const { createAutoscale } = await import("./index");
  expect(() => createAutoscale(deployment)).toThrowErrorMatchingInlineSnapshot(
    `"Missing metadata.namespace on target"`
  );
});
