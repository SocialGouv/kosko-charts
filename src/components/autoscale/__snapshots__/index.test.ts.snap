// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create an autoscale for a deployment 1`] = `
Object {
  "apiVersion": "autoscaling/v2beta2",
  "kind": "HorizontalPodAutoscaler",
  "metadata": Object {
    "name": "foo",
    "namespace": "foo-namespace",
  },
  "spec": Object {
    "maxReplicas": 10,
    "metrics": Array [
      Object {
        "resource": Object {
          "name": "cpu",
          "target": Object {
            "averageUtilization": 100,
            "type": "Utilization",
          },
        },
        "type": "Resource",
      },
      Object {
        "resource": Object {
          "name": "memory",
          "target": Object {
            "averageUtilization": 100,
            "type": "Utilization",
          },
        },
        "type": "Resource",
      },
    ],
    "minReplicas": 1,
    "scaleTargetRef": Object {
      "apiVersion": "apps/v1",
      "kind": "Deployment",
      "name": "foo",
    },
  },
}
`;

exports[`should create an autoscale with custom spec 1`] = `
Object {
  "apiVersion": "autoscaling/v2beta2",
  "kind": "HorizontalPodAutoscaler",
  "metadata": Object {
    "name": "foo",
    "namespace": "foo-namespace",
  },
  "spec": Object {
    "maxReplicas": 4,
    "metrics": Array [
      Object {
        "resource": Object {
          "name": "cpu",
          "target": Object {
            "averageUtilization": 50,
            "type": "Utilization",
          },
        },
        "type": "Resource",
      },
    ],
    "minReplicas": 2,
    "scaleTargetRef": Object {
      "apiVersion": "apps/v1",
      "kind": "Deployment",
      "name": "foo",
    },
  },
}
`;
